framework/src/util/SharedLibrary.cpp

```diff
@@ -134,36 +156,43 @@ void SharedLibrary::Load()
 void SharedLibrary::Unload()
 {
   if (d->m_Handle) {
+    if (*(d->m_Handle)) {
 #ifdef US_PLATFORM_POSIX
-    if (dlclose(d->m_Handle)) {
-      std::string err_msg = "Error unloading " + GetLibraryPath() + ".";
-      const char* err = dlerror();
-      if (err) {
-        err_msg += " " + std::string(err);
+      if (dlclose(*(d->m_Handle))) {
+        std::string err_msg = "Error unloading " + GetLibraryPath() + ".";
+        const char* err = dlerror();
+        if (err) {
+          err_msg += " " + std::string(err);
+        }
+
+        *(d->m_Handle) = nullptr;
+        throw std::runtime_error(err_msg);
       }
-
-      throw std::runtime_error(err_msg);
-    }
 #else
-    if (!FreeLibrary(reinterpret_cast<HMODULE>(d->m_Handle))) {
-      std::string errMsg = "Unloading ";
-      errMsg.append(GetLibraryPath())
-        .append("failed with error: ")
-        .append(util::GetLastWin32ErrorStr());
-
-      throw std::runtime_error(errMsg);
-    }
+      if (!FreeLibrary(reinterpret_cast<HMODULE>(*(d->m_Handle)))) {
+        std::string errMsg = "Unloading ";
+        errMsg.append(GetLibraryPath())
+          .append("failed with error: ")
+          .append(util::GetLastWin32ErrorStr());
+
+        *(d->m_Handle) = nullptr;
+        throw std::runtime_error(errMsg);
+      }
 #endif
-    d->m_Handle = nullptr;
+    }
+
+    *(d->m_Handle) = nullptr;
   }
 }
 
 void SharedLibrary::SetName(const std::string& name)
 {
   if (IsLoaded() || !d->m_FilePath.empty())
     return;
-  d.Detach();
-  d->m_Name = name;
+
+  SharedLibraryPrivate p = *d;
```

Comment:
`d` is not shared with anyone, why can't we just do `d->m_Name = name` but need to create another copy here? In fact, `make_unique` invokes the copy constructor again, effectively creating a second copy.

Same for the other changes below.